ql/src/java/org/apache/hadoop/hive/ql/io/BatchToRowReader.java

```diff
@@ -108,25 +108,41 @@ public BatchToRowReader(RecordReader<NullWritable, VectorizedRowBatch> vrbReader
     } else {
       Arrays.fill(included, true);
     }
-    // Create struct for ROW__ID virtual column and extract index
-    this.rowIdIdx = vrbCtx.findVirtualColumnNum(VirtualColumn.ROWID);
-    if (this.rowIdIdx >= 0) {
-      included[rowIdIdx] = true;
+
+    virtualColumnHandlers = requestedVirtualColumns();
+    for (VirtualColumnHandler handler : virtualColumnHandlers) {
+      int idx = vrbCtx.findVirtualColumnNum(handler.virtualColumn);
+      if (idx >= 0) {
+        included[idx] = true;
+        handler.indexInSchema = idx;
+      }
     }
+
     if (LOG.isDebugEnabled()) {
       LOG.debug("Including the columns " + DebugUtils.toString(included));
     }
     this.included = included;
   }
 
+  public static class VirtualColumnHandler {
```

Comment:
Can you add a short comment for the inner class?

Reply:
Added.